apps/astarte_pairing_api/lib/astarte_pairing_api_web/router.ex

```diff
@@ -31,6 +38,10 @@ defmodule Astarte.Pairing.APIWeb.Router do
 
     post "/agent/devices", AgentController, :create
     delete "/agent/devices/:device_id", AgentController, :delete
+  end
+
+  scope "/v1/:realm_name", Astarte.Pairing.APIWeb do
+    pipe_through :api_device
```

Comment:
For example, here too you are repeating the `/v1/:realm_name`.

However note that you did so to scope the routes that present a device ID, so that you can add `pipe_through :api_device`.
If you want to do that, then I would define a scope that matches the path until the device ID, not just the realm name.

For example, using nested scopes where each scope match the appropriate path needed to apply pipelines:

```elixir
scope "/v1/:realm_name", Astarte.Pairing.APIWeb do
  pipe_through :api
  post "/agent/devices", AgentController, :create

  scope "/devices/:hw_id", Astarte.Pairing.APIWeb do
    pipe_through :api_device
    get "/", DeviceController, :show_info
end
```

Reply:
[DeviceController](https://github.com/astarte-platform/astarte/blob/595adbd9bb042809215ae3d94c134bcf5a8ce1a1/apps/astarte_pairing_api/lib/astarte_pairing_api_web/controllers/device_controller.ex#L34C27-L34C34)  does not require plug for authorization because created in separate scope.

Reply:
As the [Phoenix docs](https://hexdocs.pm/phoenix/Phoenix.Router.html#pipeline/2) say that
> Every time [pipe_through/1](https://hexdocs.pm/phoenix/Phoenix.Router.html#pipe_through/1) is called, the new pipelines are appended to the ones previously given.

A possible solution for handling nested scopes with slightly different pipelines in this case could be
```elixir
  pipeline :api do
    plug :accepts, ["json"]
    plug Astarte.Pairing.APIWeb.Plug.LogRealm
  end

  pipeline :agent_api do
    plug Astarte.Pairing.APIWeb.Plug.AuthorizePath
  end

  pipeline :device_api do
    plug Astarte.Pairing.APIWeb.Plug.LogHwId
  end

  scope "/v1/:realm_name", Astarte.Pairing.APIWeb do
    pipe_through :api

    get "/version", VersionController, :show

    scope "/agent" do
      pipe_through :agent_api

      post "/devices", AgentController, :create
      delete "/devices/:device_id", AgentController, :delete
    end

    scope "/devices" do
      pipe_through :device_api

      get "/:hw_id", DeviceController, :show_info
      post "/:hw_id/protocols/:protocol/credentials", DeviceController, :create_credentials
      post "/:hw_id/protocols/:protocol/credentials/verify",
           DeviceController,
           :verify_credentials
    end
  end
```

What do you think @osmanhadzic @davidebriani ?

Reply:
It seems the suggestion was followed in AppEngine but not in Pairing, is there something that does not convince you @osmanhadzic?